Source/Dafny/Compilers/Compiler.cs

```diff
@@ -3068,14 +3068,15 @@ public class CheckHasNoAssumes_Visitor : BottomUpVisitor {
         var i = 0;
         foreach (var local in s.Locals) {
           bool hasRhs = s.Update is AssignSuchThatStmt;
+          bool willBeAssignedAnyway = s.Update is AssignOrReturnStmt;
           if (!hasRhs && s.Update is UpdateStmt u) {
             if (i < u.Rhss.Count && u.Rhss[i] is HavocRhs) {
               // there's no specific initial value
             } else {
               hasRhs = true;
             }
           }
-          TrLocalVar(local, !hasRhs, wr);
+          TrLocalVar(local, !hasRhs && !willBeAssignedAnyway, wr);
```

Comment:
You gave me the idea of simplifying even further:

`bool hasRhs = s.Update is AssignSuchThatStmt || s.Update is AssignOrReturnStmt;`

so that I can remove `willBeAssignedAnyway`, since it's redundant with `hasRhs`